### ISSUE TITLE
🐛 [firebase_messaging] InvalidRegistration -> NotRegistered when first notification in terminated app

### ISSUE THREAD

**Issue by saulrmzlpz** [firebase/flutterfire#11618]
## Bug report

**Describe the bug**
Doing the FCM implementation, an unusual behavior is detected when sending notifications when the application is in a terminated state on iOS.

### Steps to reproduce

1. Implement FCM according to documentation.
2. Add the following code for background management:
```dart
  @pragma('vm:entry-point')
  static Future<void> _onBackgroundMessageHandler(RemoteMessage message) async {
    Firebase.initializeApp();
    log('Handling a background message id: ${message.messageId}, data: ${message.data}', name: 'Firebase Messaging');
    print('Handling a background message id: ${message.messageId}, data: ${message.data}');
  }
```
3. Launch the app in release mode.
4. Send a test notification (I do it from postman).
5. Close the app by swiping up.
6.  Send a test notification again (This notification will arrive. We will also notice that the following message appears in the logs: [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(37)] Using the Impeller rendering backend).
7. Send a third test notification (Postman will respond with InvalidRegistration).

### Expected behavior
The expected behavior is that the token will continue to function since this behavior is not normal.

### Sample project
No sample project

---

## Additional context
info.plist value:
```xml
<key>FirebaseAppDelegateProxyEnabled</key>
<false/>
```
Xcode 15.0
iOS 17.0 iPhone 13 Pro Max

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
[✓] Flutter (Channel stable, 3.13.4, on macOS 13.5.2 22G91 darwin-x64, locale es-419)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.0)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] VS Code (version 1.82.2)
[✓] Connected device (3 available)
[✓] Network resources
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
- animated_text_kit 4.2.2 [flutter characters]
- auto_size_text 3.0.0 [flutter]
- cached_network_image 3.2.3 [flutter flutter_cache_manager octo_image cached_network_image_platform_interface cached_network_image_web]
- card_swiper 3.0.1 [flutter]
- collection 1.17.2
- core_components 0.0.16 [appinio_video_player cancellation_token_http flutter flutter_inappwebview flutter_svg fluttertoast image_gallery_saver geolocator image_picker intl lottie mime nfc_manager path path_provider photo_view share_plus sleek_circular_slider slidable_button video_compress]
- day_night_time_picker 1.3.0+1 [flutter]
- diacritic 0.1.4
- dropdown_search 5.0.6 [flutter]
- encrypt 5.0.1 [args asn1lib clock collection crypto pointycastle]
- feature_discovery 0.14.1 [flutter provider shared_preferences]
- firebase_analytics 10.5.0 [firebase_analytics_platform_interface firebase_analytics_web firebase_core firebase_core_platform_interface flutter]
- firebase_core 2.16.0 [firebase_core_platform_interface firebase_core_web flutter meta]
- firebase_crashlytics 3.3.6 [firebase_core firebase_core_platform_interface firebase_crashlytics_platform_interface flutter stack_trace]
- firebase_messaging 14.6.8 [firebase_core firebase_core_platform_interface firebase_messaging_platform_interface firebase_messaging_web flutter meta]
- flip_card 0.7.0 [flutter]
- flutter 0.0.0 [characters collection material_color_utilities meta vector_math web sky_engine]
- flutter_cache_manager 3.3.1 [clock collection file flutter http path path_provider rxdart sqflite uuid]
- flutter_inappwebview 5.7.2+3 [flutter]
- flutter_local_notifications 15.1.1 [clock flutter flutter_local_notifications_linux flutter_local_notifications_platform_interface timezone]
- flutter_localizations 0.0.0 [flutter intl characters clock collection material_color_utilities meta path vector_math web]
- flutter_mobx 2.0.6+5 [flutter mobx]
- flutter_svg 2.0.7 [flutter vector_graphics vector_graphics_codec vector_graphics_compiler]
- flutter_vibrate 1.3.0 [flutter]
- get_it 7.6.4 [async collection]
- go_router 10.1.2 [collection flutter flutter_web_plugins logging meta]
- grouped_list 5.1.2 [flutter]
- hive 2.2.3 [meta crypto]
- hive_flutter 1.1.0 [flutter hive path_provider path]
- http_visecurity 0.0.7 [flutter flutter_secure_storage http]
- in_app_update 4.2.1 [flutter]
- material_symbols_icons 4.2668.0 [flutter]
- mobile_scanner 3.4.1 [flutter flutter_web_plugins js]
- mobx 2.2.0 [meta]
- package_info_plus 4.1.0 [ffi flutter flutter_web_plugins http meta path package_info_plus_platform_interface win32]
- pdfx 2.4.0 [flutter flutter_web_plugins plugin_platform_interface js device_info_plus uuid meta extension synchronized universal_platform photo_view vector_math]
- photo_view 0.14.0 [flutter]
- provider 6.0.5 [collection flutter nested]
- qr_flutter 4.1.0 [flutter qr]
- screenshot 2.1.0 [flutter]
- shared_preferences 2.2.1 [flutter shared_preferences_android shared_preferences_foundation shared_preferences_linux shared_preferences_platform_interface shared_preferences_web shared_preferences_windows]
- shimmer 3.0.0 [flutter]
- slide_countdown 0.6.0 [flutter stream_duration]
- smooth_page_indicator 1.1.0 [flutter]
- table_calendar 3.0.9 [flutter intl simple_gesture_detector]
- url_launcher 6.1.14 [flutter url_launcher_android url_launcher_ios url_launcher_linux url_launcher_macos url_launcher_platform_interface url_launcher_web url_launcher_windows]
- video_player 2.7.1 [flutter html video_player_android video_player_avfoundation video_player_platform_interface video_player_web]
- widget_mask 1.0.0+0 [collection flutter]

dev dependencies:
- build_runner 2.4.6 [analyzer args async build build_config build_daemon build_resolvers build_runner_core code_builder collection crypto dart_style frontend_server_client glob graphs http_multi_server io js logging meta mime package_config path pool pub_semver pubspec_parse shelf shelf_web_socket stack_trace stream_transform timing watcher web_socket_channel yaml]
- flutter_gen_runner 5.3.1 [flutter_gen_core build collection crypto glob path]
- flutter_lints 2.0.3 [lints]
- flutter_test 0.0.0 [flutter test_api matcher path fake_async clock stack_trace vector_math async boolean_selector characters collection material_color_utilities meta source_span stream_channel string_scanner term_glyph web]
- hive_generator 2.0.1 [build source_gen hive analyzer source_helper]
- mobx_codegen 2.3.0 [analyzer build build_resolvers meta mobx path source_gen]

transitive dependencies:
- _fe_analyzer_shared 61.0.0 [meta]
- _flutterfire_internals 1.3.6 [collection firebase_core firebase_core_platform_interface flutter meta]
- analyzer 5.13.0 [_fe_analyzer_shared collection convert crypto glob meta package_config path pub_semver source_span watcher yaml]
- appinio_video_player 1.2.2 [video_player flutter]
- archive 3.3.8 [crypto path pointycastle]
- args 2.4.2
- asn1lib 1.5.0
- async 2.11.0 [collection meta]
- boolean_selector 2.1.1 [source_span string_scanner]
- build 2.4.1 [analyzer async convert crypto glob logging meta package_config path]
- build_config 1.1.1 [checked_yaml json_annotation path pubspec_parse yaml]
- build_daemon 4.0.0 [built_collection built_value http_multi_server logging path pool shelf shelf_web_socket stream_transform watcher web_socket_channel]
- build_resolvers 2.3.2 [analyzer async build collection convert crypto graphs logging package_config path pool pub_semver stream_transform yaml]
- build_runner_core 7.2.10 [async build build_config build_resolvers collection convert crypto glob graphs json_annotation logging meta package_config path pool timing watcher yaml]
- built_collection 5.1.1
- built_value 8.6.2 [built_collection collection fixnum meta]
- cached_network_image_platform_interface 2.0.0 [flutter flutter_cache_manager]
- cached_network_image_web 1.0.2 [flutter flutter_cache_manager cached_network_image_platform_interface]
- cancellation_token 2.0.1 [meta collection]
- cancellation_token_http 1.3.0 [async cancellation_token http_parser meta]
- characters 1.3.0
- checked_yaml 2.0.3 [json_annotation source_span yaml]
- clock 1.1.1
- code_builder 4.6.0 [built_collection built_value collection matcher meta]
- color 3.0.0
- convert 3.1.1 [typed_data]
- cross_file 0.3.3+5 [js meta]
- crypto 3.0.3 [typed_data]
- csslib 1.0.0 [source_span]
- dart_style 2.3.2 [analyzer args path pub_semver source_span]
- dartx 1.2.0 [characters collection crypto meta path time]
- dbus 0.7.8 [args ffi meta xml]
- device_info_plus 9.0.3 [device_info_plus_platform_interface ffi file flutter flutter_web_plugins meta win32 win32_registry]
- device_info_plus_platform_interface 7.0.0 [flutter meta plugin_platform_interface]
- extension 0.6.0 [meta]
- fake_async 1.3.1 [clock collection]
- ffi 2.1.0
- file 6.1.4 [meta path]
- file_selector_linux 0.9.2+1 [cross_file file_selector_platform_interface flutter]
- file_selector_macos 0.9.3+2 [cross_file file_selector_platform_interface flutter]
- file_selector_platform_interface 2.6.1 [cross_file flutter http plugin_platform_interface]
- file_selector_windows 0.9.3+1 [cross_file file_selector_platform_interface flutter]
- firebase_analytics_platform_interface 3.7.0 [_flutterfire_internals firebase_core flutter meta plugin_platform_interface]
- firebase_analytics_web 0.5.5 [_flutterfire_internals firebase_analytics_platform_interface firebase_core firebase_core_web flutter flutter_web_plugins js]
- firebase_core_platform_interface 4.8.0 [collection flutter flutter_test meta plugin_platform_interface]
- firebase_core_web 2.8.0 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- firebase_crashlytics_platform_interface 3.6.6 [_flutterfire_internals collection firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_platform_interface 4.5.7 [_flutterfire_internals firebase_core flutter meta plugin_platform_interface]
- firebase_messaging_web 3.5.7 [_flutterfire_internals firebase_core firebase_core_web firebase_messaging_platform_interface flutter flutter_web_plugins js meta]
- fixnum 1.1.0
- flutter_blurhash 0.7.0 [flutter]
- flutter_gen_core 5.3.1 [path yaml mime xml dartx color collection json_annotation glob dart_style analyzer args pub_semver source_span]
- flutter_local_notifications_linux 4.0.0+1 [dbus ffi flutter flutter_local_notifications_platform_interface path xdg_directories]
- flutter_local_notifications_platform_interface 7.0.0+1 [flutter plugin_platform_interface]
- flutter_plugin_android_lifecycle 2.0.16 [flutter]
- flutter_secure_storage 8.1.0 [flutter flutter_secure_storage_linux flutter_secure_storage_macos flutter_secure_storage_platform_interface flutter_secure_storage_web flutter_secure_storage_windows meta]
- flutter_secure_storage_linux 1.2.0 [flutter flutter_secure_storage_platform_interface]
- flutter_secure_storage_macos 3.0.1 [flutter flutter_secure_storage_platform_interface]
- flutter_secure_storage_platform_interface 1.0.2 [flutter plugin_platform_interface]
- flutter_secure_storage_web 1.1.2 [flutter flutter_secure_storage_platform_interface flutter_web_plugins js]
- flutter_secure_storage_windows 2.1.1 [flutter flutter_secure_storage_platform_interface]
- flutter_web_plugins 0.0.0 [flutter characters collection material_color_utilities meta vector_math web]
- fluttertoast 8.2.2 [flutter flutter_web_plugins]
- frontend_server_client 3.2.0 [async path]
- geolocator 9.0.2 [flutter geolocator_platform_interface geolocator_android geolocator_apple geolocator_web geolocator_windows]
- geolocator_android 4.2.2 [flutter geolocator_platform_interface uuid]
- geolocator_apple 2.2.7 [flutter geolocator_platform_interface]
- geolocator_platform_interface 4.0.8 [flutter plugin_platform_interface vector_math meta]
- geolocator_web 2.1.6 [flutter flutter_web_plugins geolocator_platform_interface]
- geolocator_windows 0.1.3 [flutter geolocator_platform_interface]
- glob 2.1.2 [async collection file path string_scanner]
- graphs 2.3.1 [collection]
- html 0.15.4 [csslib source_span]
- http 0.13.6 [async http_parser meta]
- http_multi_server 3.2.1 [async]
- http_parser 4.0.2 [collection source_span string_scanner typed_data]
- image_gallery_saver 2.0.3 [flutter]
- image_picker 1.0.4 [flutter image_picker_android image_picker_for_web image_picker_ios image_picker_linux image_picker_macos image_picker_platform_interface image_picker_windows]
- image_picker_android 0.8.7+5 [flutter flutter_plugin_android_lifecycle image_picker_platform_interface]
- image_picker_for_web 3.0.1 [flutter flutter_web_plugins image_picker_platform_interface mime]
- image_picker_ios 0.8.8+2 [flutter image_picker_platform_interface]
- image_picker_linux 0.2.1+1 [file_selector_linux file_selector_platform_interface flutter image_picker_platform_interface]
- image_picker_macos 0.2.1+1 [file_selector_macos file_selector_platform_interface flutter image_picker_platform_interface]
- image_picker_platform_interface 2.9.1 [cross_file flutter http plugin_platform_interface]
- image_picker_windows 0.2.1+1 [file_selector_platform_interface file_selector_windows flutter image_picker_platform_interface]
- intl 0.18.1 [clock meta path]
- io 1.0.4 [meta path string_scanner]
- js 0.6.7 [meta]
- json_annotation 4.8.1 [meta]
- lints 2.1.1
- logging 1.2.0
- lottie 2.6.0 [archive flutter path vector_math]
- matcher 0.12.16 [async meta stack_trace term_glyph test_api]
- material_color_utilities 0.5.0 [collection]
- meta 1.9.1
- mime 1.0.4
- nested 1.0.0 [flutter]
- nfc_manager 3.3.0 [flutter]
- octo_image 1.0.2 [flutter flutter_blurhash]
- package_config 2.1.0 [path]
- package_info_plus_platform_interface 2.0.1 [flutter meta plugin_platform_interface]
- path 1.8.3
- path_parsing 1.0.1 [vector_math meta]
- path_provider 2.1.1 [flutter path_provider_android path_provider_foundation path_provider_linux path_provider_platform_interface path_provider_windows]
- path_provider_android 2.2.0 [flutter path_provider_platform_interface]
- path_provider_foundation 2.3.1 [flutter path_provider_platform_interface]
- path_provider_linux 2.2.1 [ffi flutter path path_provider_platform_interface xdg_directories]
- path_provider_platform_interface 2.1.1 [flutter platform plugin_platform_interface]
- path_provider_windows 2.2.1 [ffi flutter path path_provider_platform_interface win32]
- petitparser 5.4.0 [meta]
- platform 3.1.2
- plugin_platform_interface 2.1.6 [meta]
- pointycastle 3.7.3 [collection convert js]
- pool 1.5.1 [async stack_trace]
- pub_semver 2.1.4 [collection meta]
- pubspec_parse 1.2.3 [checked_yaml collection json_annotation pub_semver yaml]
- qr 3.0.1 [meta]
- rxdart 0.27.7
- share_plus 7.1.0 [cross_file meta mime flutter flutter_web_plugins share_plus_platform_interface file url_launcher_web url_launcher_windows url_launcher_linux url_launcher_platform_interface ffi win32]
- share_plus_platform_interface 3.3.0 [cross_file flutter meta mime plugin_platform_interface path_provider uuid]
- shared_preferences_android 2.2.1 [flutter shared_preferences_platform_interface]
- shared_preferences_foundation 2.3.4 [flutter shared_preferences_platform_interface]
- shared_preferences_linux 2.3.1 [file flutter path path_provider_linux path_provider_platform_interface shared_preferences_platform_interface]
- shared_preferences_platform_interface 2.3.1 [flutter plugin_platform_interface]
- shared_preferences_web 2.2.1 [flutter flutter_web_plugins shared_preferences_platform_interface]
- shared_preferences_windows 2.3.1 [file flutter path path_provider_platform_interface path_provider_windows shared_preferences_platform_interface]
- shelf 1.4.1 [async collection http_parser path stack_trace stream_channel]
- shelf_web_socket 1.0.4 [shelf stream_channel web_socket_channel]
- simple_gesture_detector 0.2.0 [flutter]
- sky_engine 0.0.99
- sleek_circular_slider 2.0.1 [flutter]
- slidable_button 3.0.0+1 [flutter]
- source_gen 1.4.0 [analyzer async build dart_style glob path source_span yaml]
- source_helper 1.3.4 [analyzer collection source_gen]
- source_span 1.10.0 [collection path term_glyph]
- sqflite 2.3.0 [flutter sqflite_common path]
- sqflite_common 2.5.0 [synchronized path meta]
- stack_trace 1.11.0 [path]
- stream_channel 2.1.1 [async]
- stream_duration 3.2.1
- stream_transform 2.1.0
- string_scanner 1.2.0 [source_span]
- synchronized 3.1.0
- term_glyph 1.2.1
- test_api 0.6.0 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph]
- time 2.1.3 [clock]
- timezone 0.9.2 [path]
- timing 1.0.1 [json_annotation]
- typed_data 1.3.2 [collection]
- universal_platform 1.0.0+1
- url_launcher_android 6.1.0 [flutter url_launcher_platform_interface]
- url_launcher_ios 6.1.5 [flutter url_launcher_platform_interface]
- url_launcher_linux 3.0.6 [flutter url_launcher_platform_interface]
- url_launcher_macos 3.0.7 [flutter url_launcher_platform_interface]
- url_launcher_platform_interface 2.1.5 [flutter plugin_platform_interface]
- url_launcher_web 2.0.20 [flutter flutter_web_plugins url_launcher_platform_interface]
- url_launcher_windows 3.0.8 [flutter url_launcher_platform_interface]
- uuid 3.0.7 [crypto]
- vector_graphics 1.1.7 [flutter vector_graphics_codec]
- vector_graphics_codec 1.1.7
- vector_graphics_compiler 1.1.7 [args meta path_parsing xml vector_graphics_codec]
- vector_math 2.1.4
- video_compress 3.1.2 [flutter]
- video_player_android 2.4.10 [flutter video_player_platform_interface]
- video_player_avfoundation 2.4.10 [flutter video_player_platform_interface]
- video_player_platform_interface 6.2.1 [flutter plugin_platform_interface]
- video_player_web 2.1.0 [flutter flutter_web_plugins video_player_platform_interface]
- watcher 1.1.0 [async path]
- web 0.1.4-beta
- web_socket_channel 2.4.0 [async crypto stream_channel]
- win32 5.0.7 [ffi]
- win32_registry 1.1.1 [ffi win32]
- xdg_directories 1.0.3 [meta path]
- xml 6.3.0 [collection meta petitparser]
- yaml 3.1.2 [collection source_span string_scanner]
```

</details>

---

**Comment by darshankawar** [comment]
Thanks for the report @saulrmzlpz 


> 1. Implement FCM according to documentation.

Have you properly followed all the steps as mentioned ? Specially https://firebase.google.com/docs/cloud-messaging/flutter/client#ios ?



> 4\. Send a test notification (I do it from postman).

Since this is a third party, can you try using the plugin example's [script](https://github.com/firebase/flutterfire/blob/master/packages/firebase_messaging/firebase_messaging/example/scripts/send-message.js) and use it to send a test notification, run your scenario and check if using it, you still get the same error or not ?

Also, please provide the entire error log for further analysis.

**Comment by saulrmzlpz** [comment]
Thanks for reply.

I have reviewed and based on the flutterfire repository example reorganized the initialization and configuration methods and now it works as expected.

```dart
  static Future<void> initialize() async {
    FirebaseMessaging.onBackgroundMessage(_onBackgroundMessageHandler);
    FirebaseMessaging.onMessageOpenedApp.listen(_onMessageOpenedAppHandler);
    FirebaseMessaging.onMessage.listen(_onMessageHandler);

    await flutterLocalNotificationsPlugin
        .resolvePlatformSpecificImplementation<AndroidFlutterLocalNotificationsPlugin>()
        ?.createNotificationChannel(_channel);

    await flutterLocalNotificationsPlugin.initialize(
      const InitializationSettings(
        android: AndroidInitializationSettings('ic_stat_notification'),
        iOS: DarwinInitializationSettings(),
      ),
      onDidReceiveNotificationResponse: _receiveNotificationResponse,
    );

    await _firebaseMessaging.setForegroundNotificationPresentationOptions(
      alert: true,
      badge: true,
      sound: true,
    );

    await _requestPermission();

    getFCMToken();
  }
```

**Comment by JYeop** [comment]
Same thing happens here. 
When the app killed & receives background fcm, first message gets well.
But the second fcm not arrives, server says "Request contains an invalid argument". (Golang + firebase admin sdk)
But i looked the payload, it was totally fine everytime.
Then i launched the app, fcm token has changed. (FirebaseMessaging.instance.getToken())
It is strange bcs the token always changes when the flow happens. (app killed -> background message -> open app).
I did not  executed "deleteToken()". 

In summary, 
1. Only the first background message arrives when the app is killed, cannot receive next messages.
2. After the first background message receiving, somehow the fcm token always changed
